eventqueue/processor.go

```diff
@@ -0,0 +1,229 @@
+	isFirst := (ok && peek.Key() == r.Key()) // This is going to be true if the item being replaced is the first one in the queue
+	p.queue.Insert(r, true)
+	peek, _ = p.queue.Peek()         // No need to check for "ok" here because we know this will return an item
+	isFirst = isFirst || (peek == r) // This is also going to be true if the item just added landed at the front of the queue
+	p.process(isFirst)
+	p.queueLock.Unlock()
+
+	return nil
+}
+
+// Dequeue removes a item from the queue.
+func (p *Processor[T]) Dequeue(key string) error {
+	if p.stopped.Load() {
+		return ErrProcessorStopped
+	}
+
+	// We need to check if this is the next item in the queue, as that requires stopping the processor
+	p.queueLock.Lock()
+	peek, ok := p.queue.Peek()
+	p.queue.Remove(key)
+	if ok && peek.Key() == key {
+		// If the item was the first one in the queue, restart the processor
+		p.process(true)
+	}
+	p.queueLock.Unlock()
+
+	return nil
+}
+
+// Stop the processor.
+func (p *Processor[T]) Close() error {
+	if !p.stopped.CompareAndSwap(false, true) {
+		// Already stopped
+		return nil
+	}
+
+	// Send a signal to stop
+	close(p.stopCh)
+
+	// Blocks until process loop ends
+	t := time.NewTimer(5 * time.Second)
+	select {
+	case p.processorRunningCh <- struct{}{}:
+		// All good
+		if !t.Stop() {
+			<-t.C
+		}
+	case <-t.C:
+		return errors.New("Processor loop did not stop in 5s")
+	}
+
+	return nil
+}
+
+// Start the processing loop if it's not already running.
+// This must be invoked while the caller has a lock.
+func (p *Processor[T]) process(isNext bool) {
+	// Do not start a loop if it's already running
+	select {
+	case p.processorRunningCh <- struct{}{}:
+		// Nop - fallthrough
+	default:
+		// Already running
+		if isNext {
+			// If this is the next item, send a reset signal
+			// Use a select in case another goroutine is sending a reset signal too
+			select {
+			case p.resetCh <- struct{}{}:
+			default:
+			}
+		}
+		return
+	}
+
+	go p.processLoop()
+}
+
+// Processing loop.
+func (p *Processor[T]) processLoop() {
+	defer func() {
+		// Release the channel when exiting
+		<-p.processorRunningCh
+	}()
+
+	var (
+		r             T
+		ok            bool
+		t             kclock.Timer
+		scheduledTime time.Time
+		deadline      time.Duration
+	)
+
+	for {
+		// Continue processing items until the queue is empty
+		p.queueLock.Lock()
+		r, ok = p.queue.Peek()
+		p.queueLock.Unlock()
+		if !ok {
+			return
+		}
+
+		// Check if after obtaining the lock we have a stop or reset signals
+		// Do this before we create a timer
+		select {
+		case <-p.stopCh:
+			// Exit on stop signals
+			return
+		case <-p.resetCh:
+			// Restart the loop on reset signals
+			continue
+		default:
+			// Nop, proceed
+		}
+
+		scheduledTime = r.ScheduledTime()
+		deadline = scheduledTime.Sub(p.clock.Now())
+
+		// If the deadline is less than 0.5ms away, execute it right away
+		// This is more efficient than creating a timer
+		if deadline < 500*time.Microsecond {
```

Comment:
In the tests I've run, without this performance was suffering a lot. This is because `time.After` (or the channel from `NewTimer`) is executed after a fairly long delay, so if something needs to be done "right away", it's best to just do this.

I mean the performance was actually noticeable: tests on actor timers were failing with a timeout before I added this.